pkg/controller/bus/controller.go

```diff
@@ -309,46 +309,35 @@ func (c *Controller) syncHandler(key string) error {
 		return err
 	}
 
-	var dispatcherService *corev1.Service
-	var dispatcherDeployment, provisionerDeployment *appsv1.Deployment
-	var dispatcherServiceErr, dispatcherDeplErr, provisionerDeplError error
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	// You can use DeepCopy() to make a deep copy of original object and modify this copy
+	// Or create a copy manually for better performance
+	busCopy := bus.DeepCopy()
 
 	// Sync Service derived from the Bus
-	dispatcherService, dispatcherServiceErr = c.syncBusDispatcherService(bus)
-	if dispatcherServiceErr != nil {
-		_ = c.updateBusStatus(bus,
-			dispatcherService, dispatcherServiceErr,
-			dispatcherDeployment, dispatcherDeplErr,
-			provisionerDeployment, provisionerDeplError)
-		return dispatcherServiceErr
+	err = c.syncDispatcherServiceBusStatus(busCopy)
```

Comment:
I'm generally not a fan of passing a struct to a function expecting the function to mutate that struct. Perhaps we should split out adding status to ClusterBus and refactoring the syncHandler function.

My general thought of how to cleanup the syncHandler function is to create a status factory where we can incrementally update the resources and then call update status on that.

Something like:
```go
status := busStatusBuilder{}

// Sync Service derived from the Bus
status.dispatcherService, status.dispatcherServiceErr = c.syncBusDispatcherService(bus)
if err := status.error(); err != nil {
	_ = status.apply()
	return err
}
```

What do you think?

Reply:
My original intent was the cleanup of the `updateBusStatus` method signature but ended up splitting it into different handlers. I agree it's not nice to mutate the struct. I restored it to what I initially intended, please take a look. 

If you still think we need to break up updating the Status, the Builder approach seems reasonable to me.